keps/sig-storage/20190530-pv-health-monitor.md

```diff
@@ -34,235 +34,172 @@ status: provisional
+     - Since we want to check attaching per node in order to support multi-attach, put attaching check in node RPC here.
+  - Trigger controller and node RPC(when ready) to check local PVs health condition for local storage;
+    - For now, we do not have CSI support for local storage, we may check the local PVs directly by the agent at first, and then move the checks to RPC interfaces when ready.
 
-For networked storage, we just need to create a **controller** to check the volumes’ status periodically, and each volume plugin should implement its own health checking methods. 
-Since we are migrating in-tree volume plugins out of tree, the health checking methods should be supported in CSI.
 
-The whole architecture is as below:
-![pv health monitor architecture](./pv-health-monitor.png)
 
 
 ## Implementation
 
 ### API change
 
-We plan to introduce three taints for PV health condition.
+We plan to introduce a new Taint called PVUnhealthMessage for PV health condition whose key is specific (PVUnhealthMessage) and value can be set differently. 
 
-**One is `VolumeNotAttached` taint**,
+For example, if the PV is not attached now, we can mark the PV using the PVUnhealthMessage taint like this:
 ```
-Key: VolumeNotAttached 
-Value: true
-VolumeTaintEffect: NoEffect  (Make VolumeTaintEffect non-required field)
+Key: “PVUnhealthMessage”
+Value: “AttachError,the pv is not attached to node1 now”
+VolumeTaintEffect: NoEffect
 ```
 
-If the volume is not attached now, we need to taint the PV using the VolumeNotAttached taint. 
-
-There is already a self-healing feature in-tree that checks if a volume is still attached.  It checks every minute.  If not attached, it tries to attach again. 
-
-If the self-healing controller reattaches successfully, it should remove the VolumeNotAttached taint then. 
-
-In order not to affect the self healing feature, we set the VolumeTaintEffect to NoEffect.
-
-**Another one is `VolumeNotMounted` taint**,
+If the volume is deleted, we can mark the PV using the PVUnhealthMessage taint like this:
 ```
-Key: VolumeNotMounted 
-Value: true
-VolumeTaintEffect: NoEffect  (Make VolumeTaintEffect non-required field)
+Key: “PVUnhealthMessage” 
+Value: “VolumeError, the volume is deleted from backend”
+VolumeTaintEffect: NoEffect
 ```
 
-This can be used for both device mount and volume mount checking. If the volume is not mounted, we need to taint it.
-
-For in use volumes, if the volumes are not mounted, reactions should be taken. 
+Note that:
 
-The reactions needs further discussion, and is not in the scope of this doc. 
+- all the VolumeTaintEffects are NoEffect now at first, we may talk about the reactions later in another proposal;
+- the taint Value is string now, it is theoretically possible that several errors are detected for one PV, we may extend the string to cover this situation: combine the errors together and splited by semicolon or other symbols.
```

Comment:
If there is only one key and values are in free text form, it will be difficult to add reactions later based on those information.  Should we keep the keys such as "VolumeNotAttached" or "VolumeNotMounted" and define a few constant string values, i.e., VOLUME_DELETED, FILESYSTEM_CORRUPTED?  Maybe allow user to define their own custom values too but those will always have "NoEffect".